### ISSUE TITLE
torch.cuda.amp, example with 20% memory increase compared to apex/amp

### ISSUE THREAD

**Issue by andwurl** [pytorch/pytorch#49653]
## 🐛 Bug

I have here an example where PT1.7.1 + torch.cuda.amp requires about 20% more memory than  T1.4.0+apex/amp. What can I do to reduce the memory requirements to the level of PT1.4.0+apex/amp?

I am using a torchvision densenet121 with memory_efficient=True (gradient checkpointing enabled). 
model=models.densenet121(pretrained=None, memory_efficient=True). This would only run in PT1.7.1 native mixed precision mode with the changes mentioned here:  #40221

This model had a comparable memory footprint (2% higher) during training compared to PT1.4+apex/amp.

However, I see a significantly increased memory demand for **PT1.7.1 + native amp compared to PT1.4+ apex/amp** when I train a version of this densenet where the activation function is replaced by the following implementation of the mish activation function:

```
@torch.jit.script
def _mish_jit_fwd(x): 
    return x.mul(torch.tanh(F.softplus(x)))

@torch.jit.script
def _mish_jit_bwd(x, grad_output):
    x_sigmoid = torch.sigmoid(x)
    x_tanh_sp = F.softplus(x).tanh()
    return grad_output.mul(x_tanh_sp + x * x_sigmoid * (1 - x_tanh_sp * x_tanh_sp))

class MishJitAutoFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        return _mish_jit_fwd(x)

    @staticmethod
    def backward(ctx, grad_output):
        x = ctx.saved_variables[0]
        return _mish_jit_bwd(x, grad_output)

def mish(x): return MishJitAutoFn.apply(x)

class MishJit(nn.Module):
    @autocast()
    def forward(self, x): return MishJitAutoFn.apply(x)
```
In this case, PT1.7.1+native amp needs 20% more memory compared to PT1.4+apex/amp. Without the @autocast() in the second last line, the difference is even higher.

## Environment

PyTorch version: 1.7.1+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A
OS: SUSE Linux Enterprise Server 12 SP5 (x86_64)
GCC version: (SUSE Linux) 4.8.5
Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
Nvidia driver version: 418.39
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.7.1+cu101
[pip] torchaudio==0.7.0
[pip] torchtuples==0.2.0
[pip] torchvision==0.8.0
